File: js-old/src/views/Signer/components/util/transaction.js
// Copyright 2015-2017 Parity Technologies (UK) Ltd.
// This file is part of Parity.

// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.

import BigNumber from 'bignumber.js';

const WEI_TO_ETH_MULTIPLIER = 0.000000000000000001;
const WEI_TO_SZABU_MULTIPLIER = 0.000000000001;

export const getShortData = _getShortData;
// calculations
export const getFee = _getFee;
export const calcFeeInEth = _calcFeeInEth;
export const getTotalValue = _getTotalValue;
// displays
export const getSzaboFromWeiDisplay = _getSzaboFromWeiDisplay;
export const getValueDisplay = _getValueDisplay;
export const getValueDisplayWei = _getValueDisplayWei;
export const getTotalValueDisplay = _getTotalValueDisplay;
export const getTotalValueDisplayWei = _getTotalValueDisplayWei;
export const getEthmFromWeiDisplay = _getEthmFromWeiDisplay;
export const getGasDisplay = _getGasDisplay;

function _getShortData (data) {
  if (data.length <= 3) {
    return data;
  }
  return data.substr(0, 3) + '...';
}

/*
 * @param {hex string} gas
 * @param {wei hex string} gasPrice
 * @return {BigNumber} fee in wei
 */
function _getFee (gas, gasPrice) {
  gas = new BigNumber(gas);
  gasPrice = new BigNumber(gasPrice);
  return gasPrice.times(gas);
}

function _calcFeeInEth (totalValue, value) {
  let fee = new BigNumber(totalValue).sub(new BigNumber(value));

  return fee.times(WEI_TO_ETH_MULTIPLIER).toFormat(7);
}

/*
 * @param {wei BigNumber} fee
 * @param {wei hex string} value
 * @return {BigNumber} total value in wei
 */
function _getTotalValue (fee, value) {
  value = new BigNumber(value);
  return fee.plus(value);
}

/*
 * @param {wei hex string} gasPrice
 * @return {string} szabo gas price with unit [szabo] i.e. 21,423 [szabo]
 */
function _getSzaboFromWeiDisplay (gasPrice) {
  gasPrice = new BigNumber(gasPrice);
  return gasPrice.times(WEI_TO_SZABU_MULTIPLIER).toPrecision(5);
}

/*
 * @param {wei hex string} value
 * @return {string} value in WEI nicely formatted
 */
function _getValueDisplay (value) {
  value = new BigNumber(value);
  return value.times(WEI_TO_ETH_MULTIPLIER).toFormat(5);
}

function _getValueDisplayWei (value) {
  value = new BigNumber(value);
  return value.toFormat(0);
}

/*
 * @param {wei hex string} totalValue
 * @return {string} total value (including fee) with units i.e. 1.32 [eth]
 */
function _getTotalValueDisplay (totalValue) {
  totalValue = new BigNumber(totalValue);
  return totalValue.times(WEI_TO_ETH_MULTIPLIER).toFormat(5);
}

function _getTotalValueDisplayWei (totalValue) {
  totalValue = new BigNumber(totalValue);
  return totalValue.toFormat(0);
}

function _getEthmFromWeiDisplay (weiHexString) {
  const value = new BigNumber(weiHexString);

  return value.times(WEI_TO_ETH_MULTIPLIER).times(1e7).toFixed(5);
}

function _getGasDisplay (gas) {
  return new BigNumber(gas).times(1e-7).toFormat(4);
}

File: json/src/vm/transaction.rs
// Copyright 2015-2018 Parity Technologies (UK) Ltd.
// This file is part of Parity.

// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.

//! Executed transaction.
use hash::Address;
use uint::Uint;
use bytes::Bytes;

/// Executed transaction.
#[derive(Debug, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Transaction {
	/// Contract address.
	pub address: Address,
	/// Transaction sender.
	#[serde(rename = "caller")]
	pub sender: Address,
	/// Contract code.
	pub code: Bytes,
	/// Input data.
	pub data: Bytes,
	/// Gas.
	pub gas: Uint,
	/// Gas price.
	pub gas_price: Uint,
	/// Transaction origin.
	pub origin: Address,
	/// Sent value.
	pub value: Uint,
}

#[cfg(test)]
mod tests {
	use serde_json;
	use vm::Transaction;

	#[test]
	fn transaction_deserialization() {
		let s = r#"{
			"address" : "0f572e5295c57f15886f9b263e2f6d2d6c7b5ec6",
			"caller" : "cd1722f2947def4cf144679da39c4c32bdc35681",
			"code" : "0x7fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff7fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff01600055",
			"data" : "0x",
			"gas" : "0x0186a0",
			"gasPrice" : "0x5af3107a4000",
			"origin" : "cd1722f2947def4cf144679da39c4c32bdc35681",
			"value" : "0x0de0b6b3a7640000"
		}"#;
		let _deserialized: Transaction = serde_json::from_str(s).unwrap();
	}
}
